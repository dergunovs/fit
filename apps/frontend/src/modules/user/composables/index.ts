import { computed, Ref } from 'vue';

import { useTI18n } from '@/common/composables';

export function useUserFormTabs(isEdit: boolean, isAdmin: Ref<boolean>) {
  const { t } = useTI18n();

  const generalTab = computed(() => {
    return { value: 'general', title: t('user.general') };
  });

  const exercisesTab = computed(() => {
    return { value: 'exercises', title: t('exercise.many') };
  });

  const otherTabs = computed(() => [
    { value: 'goals', title: t('goals') },
    { value: 'weight', title: t('weight') },
    { value: 'templates', title: t('template.many') },
  ]);

  const TABS = computed(() => {
    if (!isEdit) return [generalTab.value];
    if (isAdmin.value) return [generalTab.value, ...otherTabs.value];

    return [generalTab.value, ...otherTabs.value, exercisesTab.value];
  });

  return { TABS };
}
