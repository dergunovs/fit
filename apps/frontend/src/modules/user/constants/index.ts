export const URL_USER = '/users';
export const URL_USER_CREATE = `${URL_USER}/create`;
export const URL_USER_EDIT = `${URL_USER}/edit`;
export const URL_USER_PROFILE = `${URL_USER}/profile`;

export const USER_GOALS_OPTIONS = {
  activities: [1, 2, 3, 4, 5],
  sets: [12, 16, 20, 24, 30],
  repeats: [8, 10, 12, 15, 20],
  duration: [20, 30, 40, 50, 60],
};
