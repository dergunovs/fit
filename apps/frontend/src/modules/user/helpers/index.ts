import { IEquipment, IUserEquipment } from 'fitness-tracker-contracts';

export function excludeChoosenUserEquipment(equipments: IEquipment[], userEquipments?: IUserEquipment[]) {
  return equipments.filter(
    (equipment) => !userEquipments?.some((equipmentToFilter) => equipment._id === equipmentToFilter.equipment?._id)
  );
}
