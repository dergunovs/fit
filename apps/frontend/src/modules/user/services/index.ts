import { ComputedRef, Ref } from 'vue';
import {
  API_USER,
  API_USER_PASSWORD,
  TGetUsersDTO,
  TGetUsersQueryDTO,
  TGetUserDTO,
  TPostUserDTO,
  TPostUserDataDTO,
  TUpdateUserDTO,
  TUpdateUserDataDTO,
  TDeleteUserDTO,
  TUpdateUserPasswordDTO,
  TUpdateUserPasswordDataDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const userService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_USER, page],
      queryFn: async () => {
        const params: TGetUsersQueryDTO = { page: page.value };

        const { data } = await api.get<TGetUsersDTO>(API_USER, { params });

        return data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined> | Ref<string | undefined>) => {
    return useQuery({
      queryKey: [API_USER, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetUserDTO>(`${API_USER}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_USER],
      mutationFn: async (formData: TPostUserDataDTO) => {
        const { data } = await api.post<TPostUserDTO>(API_USER, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_USER],
      mutationFn: async (formData: TUpdateUserDataDTO) => {
        const { data } = await api.patch<TUpdateUserDTO>(`${API_USER}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  updatePassword: (options: object, id?: string) => {
    return useMutation({
      mutationKey: [API_USER_PASSWORD, id],
      mutationFn: async (password: TUpdateUserPasswordDataDTO) => {
        if (!id) return null;

        const { data } = await api.patch<TUpdateUserPasswordDTO>(`${API_USER_PASSWORD}/${id}`, password);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_USER],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteUserDTO>(`${API_USER}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
