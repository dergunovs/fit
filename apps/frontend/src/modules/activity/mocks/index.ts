import { ref, computed } from 'vue';
import { vi } from 'vitest';
import { TDeleteActivityDTO, TGetActivitiesDTO } from 'fitness-tracker-contracts';

import * as activityComposables from '@/activity/composables';
import { activityService } from '@/activity/services';
import {
  ACTIVITIES_CALENDAR_FIXTURE,
  ACTIVITIES_FIXTURE,
  ACTIVITIES_STATISTICS_FIXTURE,
  ACTIVITY_FIXTURE,
  ACTIVITIES_CHART_FIXTURE,
} from '@/activity/fixtures';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { IOnSuccess } from '@/common/interface';

const dateFrom = ref('2024-11-24T21:00:00.000Z');
const dateTo = ref('2025-01-05T20:59:59.000Z');
const isDatesReady = ref(true);
const spyUpdateDates = vi.fn();

const spyGetActivity = vi.spyOn(activityService, 'getOne').mockReturnValue(mockQueryReply(ACTIVITY_FIXTURE));

const getActivitiesData: TGetActivitiesDTO = { data: ACTIVITIES_FIXTURE, total: ACTIVITIES_FIXTURE.length };

const spyGetActivities = vi
  .spyOn(activityService, 'getMany')
  .mockImplementation(() => mockQueryReply(getActivitiesData));

const spyDeleteActivity = vi.fn();

const onSuccess: IOnSuccess = {
  delete: undefined,
};

vi.spyOn(activityService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) onSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteActivityDTO, string>(spyDeleteActivity);
});

const spyUseActivityCalendar = vi.spyOn(activityComposables, 'useActivityCalendar').mockReturnValue({
  dateFrom,
  dateTo,
  isDatesReady,
  updateDates: spyUpdateDates,
});

const computedActivityCalendarEvents = computed(() =>
  activityComposables.convertActivityCalendarEvents(ACTIVITIES_CALENDAR_FIXTURE)
);

const spyComputedActivityCalendarEvents = vi
  .spyOn(activityComposables, 'computedActivityCalendarEvents')
  .mockImplementation(() => computedActivityCalendarEvents);

const spyGetActivitiesCalendar = vi
  .spyOn(activityService, 'getCalendar')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CALENDAR_FIXTURE));

const spyGetActivitiesStatistics = vi
  .spyOn(activityService, 'getStatistics')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_STATISTICS_FIXTURE));

const spyGetActivitiesChart = vi
  .spyOn(activityService, 'getChart')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CHART_FIXTURE));

export {
  spyGetActivities,
  spyGetActivity,
  spyGetActivitiesStatistics,
  spyGetActivitiesChart,
  spyDeleteActivity,
  spyUseActivityCalendar,
  spyComputedActivityCalendarEvents,
  spyGetActivitiesCalendar,
  spyUpdateDates,
  dateFrom,
  dateTo,
  isDatesReady,
  computedActivityCalendarEvents,
  getActivitiesData,
  onSuccess,
};
