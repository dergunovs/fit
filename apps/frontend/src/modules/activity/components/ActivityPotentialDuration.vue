<template>
  <p>
    <span>{{ t('activity.potentialDuration') }}: </span>

    <span v-if="props.exercises && statistics" data-test="activity-potential-duration">
      {{
        getPotentialActivityDuration(
          props.exercises,
          locale,
          statistics.exercise,
          statistics.activity.averageRestPercent.cur
        )
      }}
    </span>
  </p>
</template>

<script setup lang="ts">
import { IExerciseChoosen } from 'fitness-tracker-contracts';

import { activityService } from '@/activity/services';
import { ACTIVITY_STATISTICS_GAP } from '@/activity/constants';
import { useTI18n } from '@/common/composables';
import { getPotentialActivityDuration } from '@/activity/helpers';

interface IProps {
  exercises: IExerciseChoosen[];
}

const props = defineProps<IProps>();

const { t, locale } = useTI18n();

const { data: statistics } = activityService.getStatistics(ACTIVITY_STATISTICS_GAP);
</script>
