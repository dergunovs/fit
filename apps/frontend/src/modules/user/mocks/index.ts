import { vi } from 'vitest';
import {
  TDeleteUserDTO,
  TGetUsersDTO,
  TPostUserDataDTO,
  TPostUserDTO,
  TUpdateUserDataDTO,
  TUpdateUserDTO,
} from 'fitness-tracker-contracts';

import { userService } from '@/user/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { USER_FIXTURE, USERS_FIXTURE } from '@/user/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetUser = vi.spyOn(userService, 'getOne').mockReturnValue(mockQueryReply(USER_FIXTURE));

const mockGetUsersData: TGetUsersDTO = { data: USERS_FIXTURE, total: USERS_FIXTURE.length };

const spyGetUsers = vi.spyOn(userService, 'getMany').mockImplementation(() => mockQueryReply(mockGetUsersData));

const spyCreateUser = vi.fn();
const spyUpdateUser = vi.fn();
const spyDeleteUser = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(userService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostUserDTO, TPostUserDataDTO>(spyCreateUser);
});

vi.spyOn(userService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateUserDTO, TUpdateUserDataDTO>(spyUpdateUser);
});

vi.spyOn(userService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteUserDTO, string>(spyDeleteUser);
});

export { spyGetUser, spyGetUsers, spyCreateUser, spyUpdateUser, spyDeleteUser, mockOnSuccess, mockGetUsersData };
