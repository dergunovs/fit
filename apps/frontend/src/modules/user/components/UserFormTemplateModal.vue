<template>
  <div :class="$style.form">
    <UiFlex @submit.prevent="submit" tag="form" column gap="24" align="flex-start" data-test="template-form-modal">
      <UiField :label="t('title')" isRequired :error="error('title')">
        <UiInput v-model="formData.title" data-test="template-form-title" />
      </UiField>

      <ActivityPotentialDuration :exercises="formData.exercises" data-test="template-form-potential-duration" />

      <UiButton @click="isShowModal = true" isNarrow layout="secondary" data-test="template-form-add-exercise">
        {{ t('exercise.add') }}
      </UiButton>

      <UiModal v-model="isShowModal" isScrollable data-test="template-form-add-exercise-modal">
        <ExerciseChooseList
          v-if="exercises?.length"
          :exercises="exercises"
          @choose="addExercise"
          data-test="template-form-exercise-choose-list"
        />
      </UiModal>

      <ExerciseChoosenList
        v-if="formData.exercises?.length"
        :choosenExercises="formData.exercises"
        @delete="deleteExercise"
        @createSet="createSet"
        data-test="template-form-exercise-choosen-list"
      />

      <FormButtons
        :id="props.template?._id"
        :isEdit="!!props.template?._id"
        isHideBackButton
        @delete="(id) => emit('delete', id)"
        data-test="template-form-buttons"
      />
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { ref, onBeforeMount } from 'vue';
import { UiField, UiInput, UiFlex, UiModal, UiButton } from 'mhz-ui';
import { useValidator, required, clone, createTempId, deleteTempId } from 'mhz-helpers';
import { IExerciseChoosen, IUserTemplate } from 'fitness-tracker-contracts';

import ActivityPotentialDuration from '@/activity/components/ActivityPotentialDuration.vue';
import ExerciseChooseList from '@/exercise/components/ExerciseChooseList.vue';
import ExerciseChoosenList from '@/exercise/components/ExerciseChoosenList.vue';
import FormButtons from '@/common/components/FormButtons.vue';

import { useTI18n } from '@/common/composables';
import { exerciseService } from '@/exercise/services';

interface IProps {
  template?: IUserTemplate;
}

interface IEmit {
  create: [template: IUserTemplate];
  edit: [template: IUserTemplate];
  delete: [id: string];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();

const { t, locale } = useTI18n();

const formData = ref<IUserTemplate>({
  title: '',
  exercises: [],
});

const isShowModal = ref(false);

const { data: exercises } = exerciseService.getAll();

const { error, isValid } = useValidator(formData, { title: [required] }, locale.value);

function submit() {
  if (!isValid() || formData.value.exercises.length === 0) return;

  formData.value.exercises = deleteTempId(formData.value.exercises);

  if (props.template?._id) {
    emit('edit', formData.value);
  } else {
    emit('create', formData.value);
  }
}

function addExercise(exercise: IExerciseChoosen) {
  formData.value.exercises = formData.value.exercises?.length ? [...formData.value.exercises, exercise] : [exercise];

  isShowModal.value = false;
}

function deleteExercise(idToDelete: string) {
  formData.value.exercises = formData.value.exercises.filter((exercise) => exercise._id !== idToDelete);
}

function createSet() {
  const set = formData.value.exercises.slice(-2).map((exercise) => {
    return { ...exercise, _id: createTempId() };
  });

  formData.value.exercises = [...formData.value.exercises, ...set];
}

onBeforeMount(() => {
  if (props.template) formData.value = clone(props.template);
});
</script>

<style module lang="scss">
.form {
  min-height: 53dvh;
}
</style>
