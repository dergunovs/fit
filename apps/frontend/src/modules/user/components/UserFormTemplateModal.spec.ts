import { nextTick } from 'vue';
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest } from 'mhz-helpers';

import UserFormTemplateModal from './UserFormTemplateModal.vue';
import ExerciseChooseList from '@/exercise/components/ExerciseChooseList.vue';
import ExerciseChoosenList from '@/exercise/components/ExerciseChoosenList.vue';
import ActivityPotentialDuration from '@/activity/components/ActivityPotentialDuration.vue';
import FormButtons from '@/common/components/FormButtons.vue';

import { wrapperFactory } from '@/common/test';
import { EXERCISE_CHOOSEN_2_FIXTURE, EXERCISES_FIXTURE } from '@/exercise/fixtures';
import { spyGetExercisesAll } from '@/exercise/mocks';
import { USER_TEMPLATES } from '@/user/fixtures';

const form = dataTest('template-form-modal');
const title = dataTest('template-form-title');
const potentialDuration = dataTest('template-form-potential-duration');
const addExercise = dataTest('template-form-add-exercise');
const addExerciseModal = dataTest('template-form-add-exercise-modal');
const addExerciseChooseList = dataTest('template-form-exercise-choose-list');
const choosenExercises = dataTest('template-form-exercise-choosen-list');
const formButtons = dataTest('template-form-buttons');

let wrapper: VueWrapper<InstanceType<typeof UserFormTemplateModal>>;

const wrapperWithoutTemplate = wrapperFactory(UserFormTemplateModal, {});

beforeEach(() => {
  wrapper = wrapperFactory(UserFormTemplateModal, { template: USER_TEMPLATES[0] });
});

enableAutoUnmount(afterEach);

describe('UserFormTemplateModal', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(UserFormTemplateModal)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('gets exercises data', async () => {
    expect(spyGetExercisesAll).toBeCalledTimes(1);
  });

  it('shows potential duration', async () => {
    expect(wrapper.findComponent<typeof ActivityPotentialDuration>(potentialDuration).props('exercises')).toStrictEqual(
      USER_TEMPLATES[0].exercises
    );
  });

  it('emits edit by template submit', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('edit');

    expect(wrapper.find(addExerciseModal).attributes('modelvalue')).toStrictEqual('false');

    await wrapper.find(addExercise).trigger('click');

    expect(wrapper.find(addExerciseModal).attributes('modelvalue')).toStrictEqual('true');

    expect(wrapper.findComponent<typeof ExerciseChooseList>(addExerciseChooseList).props('exercises')).toStrictEqual(
      EXERCISES_FIXTURE
    );

    expect(wrapper.findComponent<typeof ExerciseChoosenList>(choosenExercises).props('choosenExercises')).toStrictEqual(
      USER_TEMPLATES[0].exercises
    );

    wrapper
      .findComponent<typeof ExerciseChooseList>(addExerciseChooseList)
      .vm.$emit('choose', EXERCISE_CHOOSEN_2_FIXTURE);

    await nextTick();

    expect(wrapper.find(addExerciseModal).attributes('modelvalue')).toStrictEqual('false');

    const updatedExercises = [...USER_TEMPLATES[0].exercises, EXERCISE_CHOOSEN_2_FIXTURE];

    expect(wrapper.findComponent<typeof ExerciseChoosenList>(choosenExercises).props('choosenExercises')).toStrictEqual(
      updatedExercises
    );

    await wrapper.find(form).trigger('submit');

    expect(wrapper.emitted('edit')).toHaveLength(1);
    expect(wrapper.emitted()['edit'][0]).toStrictEqual([
      { _id: USER_TEMPLATES[0]._id, title: USER_TEMPLATES[0].title, exercises: updatedExercises },
    ]);
  });

  it('emits create by template submit', async () => {
    expect(wrapperWithoutTemplate.emitted()).not.toHaveProperty('create');

    await wrapperWithoutTemplate.findComponent(title).setValue(USER_TEMPLATES[0].title);

    await wrapperWithoutTemplate.find(addExercise).trigger('click');

    wrapperWithoutTemplate
      .findComponent<typeof ExerciseChooseList>(addExerciseChooseList)
      .vm.$emit('choose', USER_TEMPLATES[0].exercises[0]);

    await wrapperWithoutTemplate.find(form).trigger('submit');

    expect(wrapperWithoutTemplate.emitted('create')).toHaveLength(1);
    expect(wrapperWithoutTemplate.emitted()['create'][0]).toStrictEqual([
      { title: USER_TEMPLATES[0].title, exercises: USER_TEMPLATES[0].exercises },
    ]);
  });

  it('emits delete by button click', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('delete');

    wrapper.findComponent<typeof FormButtons>(formButtons).vm.$emit('delete', USER_TEMPLATES[0]._id);

    await nextTick();

    expect(wrapper.emitted('delete')).toHaveLength(1);
    expect(wrapper.emitted()['delete'][0]).toStrictEqual([USER_TEMPLATES[0]._id]);
  });
});
