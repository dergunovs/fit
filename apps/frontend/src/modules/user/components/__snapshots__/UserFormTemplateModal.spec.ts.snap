// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserFormTemplateModal > matches snapshot 1`] = `
"<div class="_form_d29518">
  <div tag="form" column="" gap="24" align="flex-start" data-test="template-form-modal">
    <fieldset label="Название" isrequired=""><input type="text" modelvalue="Заголовок шаблона" data-test="template-form-title"></fieldset>
    <activity-potential-duration-stub exercises="" data-test="template-form-potential-duration"></activity-potential-duration-stub><button isnarrow="" layout="secondary" data-test="template-form-add-exercise">Добавить упражнение</button>
    <div modelvalue="false" isscrollable="" data-test="template-form-add-exercise-modal">
      <exercise-choose-list-stub exercises="[object Object],[object Object],[object Object]" data-test="template-form-exercise-choose-list"></exercise-choose-list-stub>
    </div>
    <exercise-choosen-list-stub choosenexercises="" data-test="template-form-exercise-choosen-list"></exercise-choosen-list-stub>
    <form-buttons-stub id="dgoeg1ger" isloading="false" isedit="true" isemitcancel="false" isfixed="false" ishidebackbutton="true" data-test="template-form-buttons"></form-buttons-stub>
  </div>
</div>"
`;
